File: src/main/java/de/tanschmi/aoc2024/dec1/Dec1.java
package de.tanschmi.aoc2024.dec1;

import lombok.extern.slf4j.Slf4j;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;

@Slf4j
public class Dec1 {

    public long step1(File input) throws IOException {
        ArrayList<Long> lefts = new ArrayList<>();
        ArrayList<Long> rights = new ArrayList<>();

        new ListReader().read(input, lefts, rights);

        sortColumns(lefts, rights);

        if (lefts.size() != rights.size()) {
            throw new RuntimeException("Die Länge der Liste ist unterschiedlich!!!");
        }

        long summe = calcDiffs(lefts, rights);
        log.info("Summe: {}", summe);
        return summe;
    }

    public long step2(File input) throws IOException {
        ArrayList<Long> lefts = new ArrayList<>();
        ArrayList<Long> rights = new ArrayList<>();

        new ListReader().read(input, lefts, rights);

        long sum = findSimilarities(lefts, rights);

        log.info("Summe {}", sum);
        return sum;
    }

    Long calcDiffs(ArrayList<Long> lefts, ArrayList<Long> rights) {
        long sum = 0;

        for (int i = 0; i < lefts.size(); i++) {
            long left = lefts.get(i);
            long right = rights.get(i);
            long diff;
            if (left > right) {
                diff = left - right;
            } else {
                diff = right - left;
            }
            sum += diff;
        }
        return sum;
    }

    Long findSimilarities(ArrayList<Long> lefts, ArrayList<Long> rights) {
        long scores = 0;
        for (long left : lefts) {
            long occur = rights.stream()
                    .filter(right -> right == left)
                    .count();
            long score = occur * left;
            scores += score;
        }
        return scores;
    }

    void sortColumns(ArrayList<Long> lefts, ArrayList<Long> rights) {
        sortList(lefts);
        sortList(rights);

        log.debug(lefts.toString());
        log.debug(rights.toString());
    }

    void sortList(ArrayList<Long> lefts) {

        lefts.sort(Long::compareTo);
    }
}

File: src/main/kotlin/de/tanschmi/kotlin/aoc2024/dec1/Dec1.kt
package de.tanschmi.kotlin.aoc2024.dec1

import de.tanschmi.kotlin.aoc2024.de.tanschmi.kotlin.aoc2024.FileReader
import org.apache.logging.log4j.kotlin.logger
import java.io.File

class Dec1 {
    companion object {
        private val log = logger()
    }

    fun step1(input: File): Long {

        var lefts: ArrayList<Long> = ArrayList<Long>()
        var rights: ArrayList<Long> = ArrayList<Long>()

        FileReader().readFile(input, lefts, rights)
        sortList(lefts, rights)
        var sum = calcDifference(lefts, rights)
        log.info("Summe: $sum")
        return sum
    }

    fun step2(input: File): Long {
        var lefts: ArrayList<Long> = ArrayList<Long>()
        var rights: ArrayList<Long> = ArrayList<Long>()

        FileReader().readFile(input, lefts, rights)
        sortList(lefts, rights)
        var score = calcScore(lefts, rights)
        log.info("Score: $score")
        return score
    }

    fun calcScore(lefts: ArrayList<Long>, rights: ArrayList<Long>): Long {
        var sum: Long = 0
        for (l in lefts) {
            var score: Int = rights.filter { it == l }.count()
            sum += score * l
        }
        return sum
    }

    fun calcDifference(lefts: ArrayList<Long>, rights: ArrayList<Long>): Long {
        var sum: Long = 0
        for (i in 0 until lefts.size) {
            var l = lefts[i]
            var r = rights[i]
            if (l > r) {
                sum += l - r
            } else {
                sum += r - l
            }
        }
        return sum
    }

    fun sortList(lefts: ArrayList<Long>, rights: ArrayList<Long>) {
        lefts.sort()
        rights.sort()
        log.info("Left: $lefts")
        log.info("Right: $rights")
    }



}